contracts/SourceOfRandomness.cdc

```diff
@@ -0,0 +1,20 @@
+pub contract SourceOfRandomness {
+
+    /// Canonical storage path for the SourceOfRandomness.Heartbeat resource.
+    pub let HeartbeatStoragePath: StoragePath
+
+    pub resource Heartbeat {
+        pub fun heartbeat() {
+            // called every block in the system transaction
+
+            // only callable in the system transaction and by the service account
+            let blockEntropy = entropy()
```

Comment:
So this is like `getCurrentBlock()` in that it doesn't require an import or anything?

What happens if a different account tries to call it? will it just throw a runtime error?

Reply:
> So this is like getCurrentBlock() in that it doesn't require an import or anything?

Exactly!

> What happens if a different account tries to call it? will it just throw a runtime error?

Yes. It can only be called in the system transaction actually. Otherwise it panics.

Reply:
@bluesign This is part of a proposal I'm writing a FLIP about : https://github.com/onflow/flips/pull/123/files . It's still a draft but I'm sharing an early idea of what the code changes are about. The changes will remain a draft till the FLIP is discussed and maybe approved.

Reply:
thanks @tarakby, this is a great idea (FLIP). On the implementation side, why does it have to be a function ( `entropy()` ) ? Why not send `entropy` as a parameter to heartbeat? 

I worry a bit there will be a std lib function, but not callable and it can confuse users a bit.

Reply:
> Why not send entropy as a parameter to heartbeat?

That was the other option here, but that causes some issues due to the reliance on the system TX always having the same tx hash. I would like to get rid of that problem: https://github.com/onflow/flow-go/issues/4584#issuecomment-1653578863 but I didnt want to hold back the development of this. We can later switch to arguments...

...hold on I just realised we can implement this in a way so that later when we switch to system tx arguments we don't need to make any changes to the contract... let me make that change.

Reply:
Made some changes. It should also be easier to test everything this way https://github.com/onflow/flow-go/pull/4582/files#diff-73e7c4fe793765e3190e62aecdfd6e14b10d5226f61d26a56551192e4df7d301R20